### PR TITLE
[LC-963] Run LFT as Single Node (From Genesis Invoke)

### PR DESCRIPTION
**Pull request by metalg0su**
(Possible relation: #510 )

# Goal
- Do Genesis Invoke

### CODE REVIEW

File: loopchain/consensus/runner.py [hunk 1]
```diff
@@ -86,21 +86,22 @@ async def _write_block(self, round_end_event):
...
+                #
+                # batch = db.WriteBatch()
+                # batch.put(block_hash_encoded, block_serialized)
+                # batch.put(blockchain.LAST_BLOCK_KEY, block_hash_encoded)
+                # batch.put(
+                #     blockchain.BLOCK_HEIGHT_KEY +
+                #     block.header.height.to_bytes(conf.BLOCK_HEIGHT_BYTES_LEN, byteorder='big'),
+                #     block_hash_encoded
+                # )
+                # batch.write()
```
Comment (metalg0su) [inline loopchain/consensus/runner.py:104]: Remove

File: loopchain/peer/block_manager.py [hunk 2]
```diff
@@ -868,8 +868,8 @@ def new_epoch(self):
         new_leader_id = self.get_next_leader()
         self.epoch = Epoch(self.__tx_queue, self.blockchain, new_leader_id)
         util.logger.info(f"Epoch height({self.epoch.height}), leader ({self.epoch.leader_id})")
-        if self.blockchain.block_versioner.get_version(self.epoch.height) == "1.0":
-            self.__channel_service.state_machine.start_lft()
+        # if self.blockchain.block_versioner.get_version(self.epoch.height) == "1.0":
+        #     self.__channel_service.state_machine.start_lft()
```
Comment (metalg0su) [inline loopchain/peer/block_manager.py:872]: Remove

File: loopchain/blockchain/blocks/v1_0/block.py [hunk 3]
```diff
@@ -34,6 +34,32 @@ class BlockHeader(BaseBlockHeader):
...
+        """If the block version doesn't support this, it should return None.
+        """
+        return None
+
+    @property
+    def revealed_next_reps_hash(self):
+        if self.prep_changed:
+            return self.next_validators_hash
+        return self.validators_hash
+
```
Comment (metalg0su) [inline loopchain/blockchain/blocks/v1_0/block.py:62]: Move to `BaseBlockHeader`?

File: loopchain/blockchain/blockchain.py [hunk 4]
```diff
@@ -217,6 +218,10 @@ def total_tx(self):
     def last_block(self) -> Block:
         return self.__last_block
 
+    @last_block.setter
+    def last_block(self, block):
+        self.__last_block = block
+
```
Comment (metalg0su) [inline loopchain/blockchain/blockchain.py:224]: Remove

File: loopchain/channel/channel_service.py [hunk 5]
```diff
@@ -181,6 +183,8 @@ def _manual_rollback(self) -> str:
 
     async def start_lft(self):
         self.__event_system = EventSystem()
+        self.__event_system.set_mediator(DelayedEventMediator)
+        self.__inner_service._task._event_system = self.__event_system  # --;.. ChannelInnerTask
```
Comment (hyerim-kim) [inline loopchain/channel/channel_service.py:187]: What does the comment,`--;.. ChannelInnerTask`, mean ?

File: loopchain/channel/channel_service.py [hunk 6]
```diff
@@ -203,31 +207,71 @@ async def start_lft(self):
...
-        )
-
-        # voters = self.block_manager.blockchain.find_preps_by_roothash(block.header.validators_hash)
+        invoke_pool.genesis_invoke(block)
 
         event = InitializeEvent(
-            commit_id=b'',
-            epoch_pool=[LoopchainEpoch(num=0, voters=(ChannelProperty().peer_address,))],
+            commit_id=block.header.prev_hash,
+            epoch_pool=[LoopchainEpoch(num=0, voters=()), LoopchainEpoch(num=1, voters=(ChannelProperty().peer_address,))],
```
Comment (hyerim-kim) [inline loopchain/channel/channel_service.py:221]: I think the line of 221 is too long.

File: loopchain/channel/channel_service.py (hunk 5 above)
Comment (metalg0su) [reply loopchain/channel/channel_service.py:187]: **Accidentally inserted comment.**
I found that `EventSystem` in `ChannelInnerService` should assigned AT THIS MOMENT to pass `IF` phrase in somewhere. So I write temporary comment to fix it.

I will remove that soon.

File: loopchain/channel/channel_service.py (hunk 6 above)
Comment (metalg0su) [reply loopchain/channel/channel_service.py]: Shortened. Thank you.

File: loopchain/channel/channel_service.py (hunk 5 above)
Comment (metalg0su) [reply loopchain/channel/channel_service.py]: Replaced to a more meaningful comment.

File: loopchain/channel/channel_service.py [hunk 7]
```diff
@@ -203,31 +207,71 @@ async def start_lft(self):
...
+        genesis_data_path = conf.CHANNEL_OPTION[ChannelProperty().name]["genesis_data_path"]
+        utils.logger.spam(f"Try to load a file of initial genesis block from ({genesis_data_path})")
+        with open(genesis_data_path, encoding="utf-8") as json_file:
+            tx_info = json.load(json_file)["transaction_data"]
+            nid = tx_info["nid"]
+            ChannelProperty().nid = nid
+
+        tx_builder = TransactionBuilder.new("genesis", "", tx_versioner)
+        nid = tx_info.get("nid")
+        if nid is not None:
```
Comment (hyerim-kim) [inline loopchain/channel/channel_service.py:271]:
```suggestion
        tx_builder = TransactionBuilder.new("genesis", "", tx_versioner)
        nid = tx_info.get("nid")
        tx_builder.nid = int(nid, 16) if nid else None
```

Comment (metalg0su) [reply loopchain/channel/channel_service.py]: Changes applied! 
Thank you.

File: loopchain/consensus/runner.py [hunk 8]
```diff
@@ -18,33 +30,83 @@ def __init__(self,
...
+    # FIXME: Temporary
+    async def _write_block(self, round_end_event):
+        utils.logger.notice(f"> EPOCH // ROUND ({round_end_event.epoch_num} // {round_end_event.round_num})")
+        utils.logger.notice(f"> Candidate id: {round_end_event.candidate_id}")
+        utils.logger.notice(f"> Commit id: {round_end_event.commit_id}")
 
-    async def _on_event_receive_vote(self, vote: 'Vote'):
-        await self.consensus.receive_vote(vote)
+        if round_end_event.is_success:
+            consensus_db_pool = self.consensus._data_pool  # FIXME
```
Comment (metalg0su) [inline loopchain/consensus/runner.py:80]: Use `CandidateBlocks`

File: loopchain/consensus/runner.py [hunk 9]
```diff
@@ -18,33 +30,83 @@ def __init__(self,
...
+        await self._write_block(round_end_event)
+        await self._round_start(round_end_event)
 
-    async def _on_event_receive_data(self, block: 'Data'):
-        await self.consensus.receive_data(block)
+    # FIXME: Temporary
+    async def _write_block(self, round_end_event):
+        utils.logger.notice(f"> EPOCH // ROUND ({round_end_event.epoch_num} // {round_end_event.round_num})")
+        utils.logger.notice(f"> Candidate id: {round_end_event.candidate_id}")
+        utils.logger.notice(f"> Commit id: {round_end_event.commit_id}")
```
Comment (metalg0su) [inline loopchain/consensus/runner.py:77]: Check that `BlockChain.add_block` prints similar logs.

File: loopchain/channel/channel_service.py [hunk 10]
```diff
@@ -181,6 +183,9 @@ def _manual_rollback(self) -> str:
 
     async def start_lft(self):
         self.__event_system = EventSystem()
+        self.__event_system.set_mediator(DelayedEventMediator)
+        # FIXME: How to initialize ChannelInnerService._event_system better way?
+        self.__inner_service._task._event_system = self.__event_system
 
         epoch_pool = EpochPool()
```
Comment (metalg0su) [inline loopchain/channel/channel_service.py:190]: Check its usage.

File: loopchain/consensus/runner.py (hunk 8 above)
Comment (hyerim-kim) [inline loopchain/consensus/runner.py:80]: How about using `CandidateBlocks` in `BlockManager` for getting a candidate block to commit? Accessing the data pool in `LFT` by the `loopchain` looks weird.

File: loopchain/consensus/runner.py [hunk 11]
```diff
@@ -18,33 +30,83 @@ def __init__(self,
...
 
-    async def _on_event_receive_vote(self, vote: 'Vote'):
-        await self.consensus.receive_vote(vote)
+        if round_end_event.is_success:
+            consensus_db_pool = self.consensus._data_pool  # FIXME
+            blockchain = self._block_manager.blockchain
+            db = blockchain.blockchain_store
```
Comment (hyerim-kim) [inline loopchain/consensus/runner.py:82]: It's never been used.

File: loopchain/consensus/runner.py [hunk 12]
```diff
@@ -18,33 +30,83 @@ def __init__(self,
...
     def start(self, event: InitializeEvent):
         self.event_system.start(blocking=False)
         self.event_system.simulator.raise_event(event)
 
-    async def _on_event_broadcast_data(self, block: 'Data'):
-        # call broadcast block
-        pass
+    async def _on_event_broadcast_data(self, event: BroadcastDataEvent):
+        target_reps_hash = ChannelProperty().crep_root_hash  # FIXME
+        self._block_manager._send_unconfirmed_block(
```
[the remark this answers is not in the document]
Comment (metalg0su) [reply loopchain/consensus/runner.py]: `send_unconfirmed_block` exposed.

File: loopchain/consensus/runner.py [hunk 13]
```diff
@@ -1,15 +1,27 @@
...
 from typing import TYPE_CHECKING
 
 from lft.consensus import Consensus
-from lft.consensus.events import BroadcastDataEvent, BroadcastVoteEvent, ReceiveDataEvent, ReceiveVoteEvent, \
-    InitializeEvent
+from lft.consensus.events import BroadcastDataEvent, BroadcastVoteEvent, InitializeEvent, RoundEndEvent, RoundStartEvent
 from lft.event import EventSystem, EventRegister
+from lft.event.mediators import DelayedEventMediator
+
+from loopchain import configure_default as conf
```
[the remark this answers is not in the document]
Comment (metalg0su) [reply loopchain/consensus/runner.py]: Removed!

File: loopchain/consensus/runner.py (hunk 11 above)
Comment (metalg0su) [reply loopchain/consensus/runner.py]: Removed!

File: loopchain/consensus/runner.py (hunk 9 above)
Comment (metalg0su) [reply loopchain/consensus/runner.py]: Removed.

File: loopchain/consensus/runner.py (hunk 8 above)
Comment (metalg0su) [reply loopchain/consensus/runner.py:75]: Duplicated: https://github.com/icon-project/loopchain/pull/517#discussion_r442695779

Comment (metalg0su) [reply loopchain/consensus/runner.py:75]: > Accessing the data pool in LFT by the loopchain looks weird.
I totally agree.

LFT supplies ONLY `commit_id` to application, not the commited `Block`...
which means that Loopchain as application should access candidate blocks by using `commit_id`.

We have one of followings options:
- Use `CandidateBlocks` to get commited block.
- Use shared `DataPool` between app and LFT (currently, LFT has its own `DataPool`, initialized at the first moment of `Consensus`)
- Pass committed block with `round_end_event` (I'm afraid this makes lots of side effects or major changes.)

Any ideas?

File: loopchain/blockchain/blocks/v1_0/block.py (hunk 3 above)
Comment (windies21) [reply loopchain/blockchain/blocks/v1_0/block.py:62]: Have you tried it?

Comment (windies21) [reply loopchain/blockchain/blocks/v1_0/block.py:62]: I think they look common but not common.

File: loopchain/channel/channel_service.py (hunk 10 above)
Comment (windies21) [reply loopchain/channel/channel_service.py:190]: How about making an event_system setter?

File: loopchain/peer/block_manager.py (hunk 2 above)
Comment (windies21) [reply loopchain/peer/block_manager.py:872]: We will fix it in another PR.

File: loopchain/consensus/runner.py (hunk 8 above)
Comment (windies21) [reply loopchain/consensus/runner.py:75]: I think it's enough now, but We'll have to fix it later. Basically 'Use shared DataPool' is better.

**Pull request merged**